A59-audit-logging.md

```diff
@@ -0,0 +1,243 @@
+  //
+  AuditCondition audit_condition = 3;
+}
+```
+
+Note that `DENY` and `ALLOW` in the enum refer to the RBAC decisions, not
+RBAC actions.
+
+For the audit logger configuration, we will follow the existing `access_log`
+field in the [HTTP Connection Manager][HttpConnectionManager proto] by adding
+a typed extension category `audit_log`:
+
+```
+package envoy.extensions.filters.network.http_connection_manager.v3;
+
+message HttpConnectionManager {
+  ...
+  
+  // Configuration for the audit log.
+  // The audit_loggers instantiated from this will be used in all RBAC filters.
+  //
+  // [#extension-category: envoy.audit_loggers]
+  repeated xds.core.v3.TypedExtensionConfig audit_log = 53;
+}
+```
+
+The list of configured audit loggers will all be invoked by an RBAC filter when
+the audit condition is met. This is analogus to the semantics that all access
+loggers will be run after the response is sent.
+
+### gRPC Authorzation Policy Changes
+
+The gRPC authorization policy consists of `deny_rules` and `allow_rules`, which
+are mapped to two RBACs. Overall it's a subset of the RBAC as stated in [A43: gRPC authorization API][A43].
+We will keep this philosophy by adding just one `audit_condition` in the policy
+instead of having separate conditions for those two set of rules. Likewise, only
+one audit logger can be specified, the configuration of which is the json
+representation of the typed config used in xDS.
+
+Following are changes to the JSON schema of gRPC Authorization Policy introduced
+in [A43: gRPC authorization API][A43]:
+
+```json
+{
+  "title": "AuthorizationPolicy",
+  "definitions": {
+    ...
+    "audit_log": {
+      "description": "Configuration for the audit log.",
+      "type": "object",
+      "properties": {
+        "name": {
+          "description": "The name of the audit_log configuration."
+            "It is mainly used for human purposes.",
+          "type": "string",
+        },
+        "typed_config": {
+          "description": "The typed config for the audit_log."
+            "This needs to be a json object mapped from google.protobuf.Any"
+            "proto message.",
+          "type": "object",
+        }
+      },
+    }
+  },
+  "properties": {
+    ...
+    "audit_condition": {
+      "description": "Condition for the audit logging to be happen."
+        "Same as NONE if ommited.",
+      "type": "string",
+      "enum": ["NONE", "ON_DENY", "ON_ALLOW", "ON_DENY_AND_ALLOW"]
+    },
+    "audit_log": {
+      "$ref": "#/definitions/audit_log"
+    }
+  },
+  "required": ["name", "allow_rules"]
+}
+```
+
+For readers that are more familiar with proto messages, the change translates
+to:
+
+```
+message AuthorizationPolicy {
+  …
+
+  enum AuditCondition {
+    NONE = 0;
+    ON_DENY = 1;
+    ON_ALLOW = 2;
+    ALWAYS = 3;
+  }
+
+  AuditCondition audit_condition = 1;
+
+  // See https://github.com/cncf/xds/blob/main/xds/core/v3/extension.proto.
+  xds.core.v3.TypedExtensionConfig audit_log = 2;
```

Comment:
I don't think we should use the xDS `TypedExtensionConfig` message here.  Instead, I think we should do something like this:

```proto3
message AuditLogger {
  string name = 1;
  google.protobuf.Struct config = 2;  
}
AuditLogger audit_logger = 2;
```

Reply:
Done. Note I used `AuditLog` and `audit_log` instead of `...logger` to be consistent with the xDS RBAC change which is to be consistent with `access_log` field.

Reply:
I don't think we need to be consistent with xDS here, since this is a different API.  I think it makes more sense to call it `AuditLogger` and to call the field `audit_logger`.

Reply:
Makes sense! Renamed the field.